srml/im-online/src/lib.rs

```diff
@@ -233,27 +233,57 @@ decl_module! {
 						.ok_or(OffchainErr::ExtrinsicCreation)?;
 					sr_io::submit_transaction(&ex)
 						.map_err(|_| OffchainErr::SubmitTransaction)?;
+
+					// once finished we set the worker status without comparing
+					// if the existing value changed in the meantime. this is
+					// because at this point the heartbeat was definitely submitted.
 					set_worker_status::<T>(block_number, true);
 				}
 				Ok(())
 			}
 
-			fn set_worker_status<T: Trait>(gossipping_at: T::BlockNumber, done: bool) {
+			fn compare_and_set_worker_status<T: Trait>(
+				gossipping_at: T::BlockNumber,
+				done: bool,
+				curr_worker_status: Option<Vec<u8>>,
+			) -> bool {
+				let enc = WorkerStatus {
+					done,
+					gossipping_at,
+				};
+				match curr_worker_status {
```

Comment:
Is it necessary to use `match`? Can't we do `curr_worker_status.map(Vec::as_slice)`?

Reply:
Great idea, I updated the PR.